docs/mimir.1
.\" Automatically generated by Pandoc 1.19.2.4
.\"
.TH "MIMIR" "1" "February 2021" "mimir 0.3.0" "Usage manual"
.hy
.SH NAME
.PP
\f[C]mimir\f[] \-\- an optimization\-based purse seine path planner
.SH SYNOPSIS
.PP
\f[C]/opt/sintef/bin/mimir\ [OPTIONS]\ CONFIG\f[]
.SH DESCRIPTION
.PP
A purse seine path planner algorithm that calculates suggested
deployment paths based on user preferences and environmental conditions.
.SH OPTIONS
.PP
The command is valid with install prefix set to
\f[I]\f[C]/opt/sintef\f[]\f[].
If \f[C]$PATH\f[] has \f[C]/opt/sintef/bin\f[], the executable name is
simply \f[C]mimir\f[].
.TP
.B \f[C]CONFIG\f[]
Path to YAML config file.
.RS
.RE
.TP
.B \f[C]\-h,\ \-\-help\f[]
Display the help message.
.RS
.RE
.TP
.B \f[C]\-v\ number,\ \-\-severity\ number\f[]
Print severity, with \f[I]number\f[] as defined in \f[I]Severity\f[]
below, default: 3.
.RS
.RE
.TP
.B \f[C]\-l\ number,\ \-\-log\ number\f[]
Log file severity: with \f[I]number\f[] as defined in \f[I]Severity\f[]
below, default: 3.
.RS
.RE
.PP
\f[I]Severity\f[] is an integer \f[I]number\f[] in [0\-5], 5: trace, 4:
debug, 3: info, 2: warning, 1: error, 0: fatal
.SH CONFIGURATION
.PP
The YAML configuration file follows a specific schema, with two common
map attributes: \f[I]dds\f[] and \f[I]algorithm\f[].
The \f[I]algorithm:name\f[] defines the name of the algorithm to be run
and imples that a map with the same name is present in the YAML file.
The configuration specification for a selected algorithm is specific to
that particular algorithm.
See \f[I]EXAMPLE\f[] below.
.SH ENVIRONMENT
.TP
.B \f[C]OSPL_URI\f[]
If \f[C]$OSPL_URI\f[] is not set, it defaults to
\f[I]\f[C]file:///opt/sintef/etc/config/ospl.xml\f[]\f[].
This is the configuration file for the OpenSplice middleware service
used by the application.
See \f[B]OpenSplice deployment\f[] for details on configuring the
middleware service.
.RS
.RE
.TP
.B \f[C]OSPL_HOME\f[]
If \f[C]$OSPL_HOME\f[] is not set, it defaults to
\f[I]\f[C]/opt/sintef/\f[]\f[].
This is the root path used by the OpenSplice middleware.
.RS
.RE
.SH FILES
.TP
.B \f[I]\f[C]/opt/sintef/etc/config/ospl.xml/\f[]\f[]
OpenSplice configuration file as described in \f[B]OpenSplice
deployment\f[].
.RS
.RE
.SH EXAMPLE
.TP
.B \f[C]/opt/sintef/bin/mimir\ \-v\ 0\ \-l\ 3\ /path/to/config.yml\f[]
Run algorithm specified in \f[I]\f[C]/path/to/config.yml/\f[]\f[] with
fatal severity output to terminal, and info severity output to log file.
The YAML config file below shows the common schema that \f[C]mimir\f[]
expects.
Note that the algoritm enters a standby state and will not execute
before a \f[C]START_PROCESS\f[] command is issued to the
\f[C]request_topic\f[] specified in the config file.
.RS
.RE
.SS YAML config file
.PP
The \f[C]algorithm:command\f[] and \f[C]algorithm:notifier\f[] entries
are used to configure the DDS topics for user commands and state machine
transition notifications.
They use the interface definition language types indicated in the
comment, which are defined in the \f[C]RatatoskIDL\f[] dependency.
Each algorithm has its own \f[I]algorithm\-specific\f[] schema; consult
the documentation for the algorithm you intend to use.
.IP
.nf
\f[C]
#\ Excerpt\ of\ /path/to/config.yml,\ with\ \[aq]algorithm\-specific\[aq]\ map\ not\ expanded:
\-\-\-
dds:
\ \ domain:\ 0\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ #\ DDS\ domain\ to\ connect\ to
algorithm:
\ \ name:\ PursePlanner\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ #\ Name\ identifier\ of\ algorithm
\ \ command:
\ \ \ \ request_topic:\ fkin_cmd\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ #\ IDL\ type:\ fkin::Command
\ \ \ \ reply_topic:\ fkin_cmd_resp\ \ \ \ \ \ \ \ \ \ \ \ \ #\ IDL\ type:\ fkin::CommandResponse
\ \ \ \ recipient:\ PursePlanner\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ #\ Recipient\ identifier\ request/reply
\ \ notifier:
\ \ \ \ notify_topic:\ fkin_state_notification\ \ #\ IDL\ type:\ fkin::ProcessStateKind
\ \ \ \ identifier:\ PursePlanner\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ #\ Identifier\ of\ notification

PursePlanner:
\ \ *algorithm\-specific
\&...
\f[]
.fi
.SH NOTES
.PP
Currently, the log outputs are only printed to stdout.
This means that the \f[B]\f[C]\-l\ |\ \-\-log\f[]\f[] option has no
effect.
.SH SEE ALSO
.IP \[bu] 2
\f[B]yamllint\f[].1 can be useful to validate the YAML config file (not
installed by default).
.IP \[bu] 2
OpenSplice
deployment (http://download.prismtech.com/docs/Vortex/html/ospl/DeploymentGuide/guide.html)
.SH COPYRIGHT
.PP
Copyright © 2021 SINTEF Ocean AS.
License: Apache\-2.0 (http://www.apache.org/licenses/LICENSE-2.0).
.SH AUTHORS
Joakim Haugen (joakim.haugen\@sintef.no).


src/programs/mimir.cpp
#include <atomic>
#include <cstdlib>
#include <cinttypes>
#include <chrono>
#include <exception>
#include <filesystem>
#include <functional>
#include <future>
#include <iostream>
#include <string>
#include <memory>
#include <thread>

#include <yaml-cpp/yaml.h>

#include "mimir/mimir.hpp"
#include "mimir/program/Convenience.hpp"
#include "mimir/program/Options.hpp"
#include "mimir/StateMachine.hpp"

#if defined(_WIN32) || defined(_WIN64)
int setenv(const char *name, const char *value, int overwrite)
{
  int errcode = 0;
  if(!overwrite) {
    size_t envsize = 0;
    errcode = getenv_s(&envsize, NULL, 0, name);
    if(errcode || envsize) return errcode;
  }
  return _putenv_s(name, value);
}

#include <windows.h>
#include <stdio.h>
#include <signal.h>

  BOOL CtrlHandler(DWORD fdwCtrlType)
  {
    switch (fdwCtrlType)
    {
      // Handle the CTRL-C signal.
    case CTRL_C_EVENT:
      mimir::signaled = 1;
      return TRUE;
      // Pass other signals to the next handler.
    case CTRL_BREAK_EVENT:
      mimir::signaled = 1;
      return TRUE;
    default:
      return FALSE;
    }
  }
#endif

int main(int argc, char *argv[])
{

  std::signal(SIGTERM, mimir::SIGTERMHandler);
  std::signal(SIGINT, mimir::SIGTERMHandler);
#if defined(_WIN32) || defined(_WIN64)
  SetConsoleCtrlHandler((PHANDLER_ROUTINE) CtrlHandler, TRUE);
#endif

  mimir::Options options("Purse seine state estimator");
  YAML::Node config;

  // Load configuration
  try
  {
    if(!options.parse(argc, argv))
      return EXIT_FAILURE;

    boost::log::add_console_log(std::cout, boost::log::keywords::auto_flush = true);
    boost::log::core::get()->set_filter(
        boost::log::trivial::severity >=
        mimir::transform_severity(options.severity) );

    config = YAML::LoadFile(options.filename); // load the file with YAML, (future: check schema with YAVL)

#ifdef MIMIR_INSTALL_PREFIX
    // This is not relocatable, but works if installed in default location provided by CPack installer
    int do_overwrite = 0;
    setenv("OSPL_HOME", MIMIR_INSTALL_PREFIX, do_overwrite);
    BOOST_LOG_TRIVIAL(debug) << "OSPL_HOME: " << getenv("OSPL_HOME") << std::endl;
    char prefix[] = MIMIR_INSTALL_PREFIX;
    namespace fs = std::filesystem;
    fs::path ospl_path = "file://";
    ospl_path += prefix;
    ospl_path /= "etc/config/ospl.xml";
    setenv("OSPL_URI", ospl_path.string().c_str(), do_overwrite);
    BOOST_LOG_TRIVIAL(debug) << "OSPL_URI: " << getenv("OSPL_URI") << std::endl;
#endif

    std::string algorithm = config["algorithm"]["name"].as<std::string>();

    if(algorithm.compare("PursePlanner") == 0)
      BOOST_LOG_TRIVIAL(info) << mimir::PlannerBanner();
    else if(algorithm.compare("Leadline") == 0)
      BOOST_LOG_TRIVIAL(info) << mimir::LeadlineBanner();
    else if(algorithm.compare("FishSchool") == 0)
      BOOST_LOG_TRIVIAL(info) << mimir::FishSchoolBanner();
    else if(algorithm.compare("KinematicVessel") == 0)
      BOOST_LOG_TRIVIAL(info) << mimir::KinematicVesselBanner();
  }
  catch (boost::program_options::error& e)
  {
    BOOST_LOG_TRIVIAL(fatal) << e.what();
    return EXIT_FAILURE;
  }
  catch (const YAML::ParserException& e)
  {
    BOOST_LOG_TRIVIAL(fatal) << "YAML parser exception: " + std::string(e.what());
    return EXIT_FAILURE;
  }

  std::thread stateMachineThread;
  int status = EXIT_SUCCESS;
  // Start main execution
  try
  {

    std::atomic<bool> machine_destructed(false);
    mimir::bsc::fifo_scheduler<> scheduler(true);
    mimir::bsc::fifo_scheduler<>::processor_handle stateMachine;
    std::promise<void> the_promise;

    stateMachine = scheduler.create_processor<mimir::StateMachine>(config, std::ref(machine_destructed));
    scheduler.initiate_processor(stateMachine);

    // if autostart:
    //    scheduler.queue_event(stateMachine, make_intrusive(new mimir::EvStart()));

    stateMachineThread = std::thread(
        [&the_promise, &scheduler, &machine_destructed]
        {
          try
          {
            scheduler(0);
            the_promise.set_value();
          }
          catch (...)
          {
            machine_destructed = true;
            the_promise.set_exception(std::current_exception());
          }
        });

    std::future<void> the_future = the_promise.get_future();

    while(!mimir::signaled && !machine_destructed)
    {
      std::this_thread::sleep_for(std::chrono::milliseconds(1000));
      if(mimir::signaled){ BOOST_LOG_TRIVIAL(info) << "Ctrl-C"; }
    }

    if (mimir::signaled)
    {
      BOOST_LOG_TRIVIAL(info) << "Posting Kill command to State Machine";
      scheduler.queue_event(stateMachine, make_intrusive(new mimir::EvKill()));
    }

    the_future.get(); // handles exceptions

  }
  catch (const YAML::Exception& e)
  {
    BOOST_LOG_TRIVIAL(fatal) << "YAML general exception: " + std::string(e.what());
    status = EXIT_FAILURE;
  }
  catch (const std::runtime_error& e)
  {
    BOOST_LOG_TRIVIAL(fatal) << "std::runtime_error: " + std::string(e.what());
    status = EXIT_FAILURE;
  }
  catch (const std::exception& e) {
    BOOST_LOG_TRIVIAL(fatal) << "std::exception: " + std::string(e.what());
    status = EXIT_FAILURE;
  }
  catch (...) {
    BOOST_LOG_TRIVIAL(fatal) << "Unknown error!";
    status = EXIT_FAILURE;
  }

  BOOST_LOG_TRIVIAL(debug) << "Waiting for StateMachine thread destruction";
  stateMachineThread.join();
  BOOST_LOG_TRIVIAL(debug) << "Program is shutting down";
  return status;

}
